db/migrations/V2.14.0__Add_unique_constraints.sql

```diff
@@ -0,0 +1,82 @@
+    (kunta_koodi)
+    WHERE katu.katunimi_fi is null and katu.katunimi_sv is null;
+
+-- All osoite fields are nullable. Most addresses have all fields.
+-- Is not distinct from is too slow, better delete null and non-null
+-- duplicates using separate queries.
+DELETE from jkr.osoite duplicate
+    WHERE exists (
+        select 1 from jkr.osoite where
+            osoite.osoitenumero = duplicate.osoitenumero and
+            osoite.katu_id = duplicate.katu_id and
+            osoite.rakennus_id = duplicate.rakennus_id and
+            osoite.posti_numero = duplicate.posti_numero
+            and osoite.id < duplicate.id
+    );
+CREATE UNIQUE INDEX idx_osoite_osoitenumero_katu_id_rakennus_id_posti_numero
+    ON jkr.osoite
+    (osoitenumero, katu_id, rakennus_id, posti_numero)
+    WHERE osoitenumero is not null;
+
+-- Some addresses are missing osoitenumero.
+DELETE from jkr.osoite duplicate
+    WHERE exists (
+        select 1 from jkr.osoite where
+            osoite.osoitenumero is null and duplicate.osoitenumero is null and
+            osoite.katu_id = duplicate.katu_id and
+            osoite.rakennus_id = duplicate.rakennus_id and
+            osoite.posti_numero = duplicate.posti_numero
+            and osoite.id < duplicate.id
+    );
+CREATE UNIQUE INDEX idx_osoite_katu_id_rakennus_id_posti_numero
+    ON jkr.osoite
+    (katu_id, rakennus_id, posti_numero)
+    WHERE osoitenumero is null and katu_id is not null and posti_numero is not null;
+
+-- Some addresses are missing katu_id and only have postinumero.
+DELETE from jkr.osoite duplicate
+    WHERE exists (
+        select 1 from jkr.osoite where
+            osoite.osoitenumero is null and duplicate.osoitenumero is null and
+            osoite.katu_id is null and duplicate.katu_id is null and
+            osoite.rakennus_id = duplicate.rakennus_id and
+            osoite.posti_numero = duplicate.posti_numero
+            and osoite.id < duplicate.id
+    );
+CREATE UNIQUE INDEX idx_osoite_rakennus_id_posti_numero
+    ON jkr.osoite
+    (rakennus_id, posti_numero)
+    WHERE osoitenumero is null and katu_id is null and posti_numero is not null;
+
+-- Some addresses might only have kunta_id. In this case, only katu_id will exist,
+-- pointing to an empty street.
+DELETE from jkr.osoite duplicate
+    WHERE exists (
+        select 1 from jkr.osoite where
+            osoite.osoitenumero is null and duplicate.osoitenumero is null and
+            osoite.katu_id = duplicate.katu_id and
+            osoite.rakennus_id = duplicate.rakennus_id and
+            osoite.posti_numero is null and duplicate.posti_numero is null
+            and osoite.id < duplicate.id
+    );
+CREATE UNIQUE INDEX idx_osoite_katu_id_rakennus_id
+    ON jkr.osoite
+    (katu_id, rakennus_id)
+    WHERE osoitenumero is null and katu_id is not null and posti_numero is null;
+
+DROP INDEX idx_osapuoli_henkilotunnus;
+CREATE UNIQUE INDEX idx_osapuoli_henkilotunnus ON jkr.osapuoli USING btree (henkilotunnus);
```

Comment:
Here we need also to clean the data before enforcing uniqueness on hetu and y-tunnus. And in addition to delete we need to update references of those that points to records we are going to delete 😰.
```sql
UPDATE jkr.kohteen_osapuolet
SET osapuoli_id = osapuoli.id_to_keep
FROM (
  SELECT
    id,
    first_value(id) OVER (PARTITION BY henkilotunnus ORDER BY id) AS id_to_keep
  FROM osapuoli
) AS osapuoli
WHERE
  osapuoli.id = kohteen_osapuolet.osapuoli_id
  AND osapuoli.id <> osapuoli.id_to_keep;
```
And same to `rakennuksen_omistajat` and `kiinteiston_omistajat`.

Reply:
Changed unique index to allow multiple identical ytunnus for non-dvv osapuolet.

Didn't touch henkilötunnus index, assuming there are no multiple osapuolet with same henkilötunnus.